funblocks-server/src/Funblocks.hs

```diff
@@ -0,0 +1,261 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE CPP #-}
+{-
+  Copyright 2017 The CodeWorld Authors. All rights reserved.
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+      http://www.apache.org/licenses/LICENSE-2.0
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+-}
+
+module Funblocks (funblockRoutes, ClientId(..)) where
+
+import           Control.Monad
+import           Control.Monad.Trans
+import           Data.Aeson
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Char8 as BC
+import qualified Data.ByteString.Lazy as LB
+import           Data.List
+import           Data.Maybe
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+import           Network.HTTP.Conduit
+import           Snap.Core
+import           Snap.Util.FileServe
+import           System.Directory
+import           System.FilePath
+
+import Model
+import Util
+
+newtype ClientId = ClientId (Maybe T.Text) deriving (Eq)
+
+-- Retrieves the user for the current request.  The request should have an
+-- id_token parameter with an id token retrieved from the Google
+-- authentication API.  The user is returned if the id token is valid.
+getUser :: ClientId -> Snap User
+getUser clientId = getParam "id_token" >>= \ case
+    Nothing       -> pass
+    Just id_token -> do
+        let url = "https://www.googleapis.com/oauth2/v1/tokeninfo?id_token=" ++ BC.unpack id_token
+        decoded <- fmap decode $ liftIO $ simpleHttp url
+        case decoded of
+            Nothing -> pass
+            Just user -> do
+                when (clientId /= ClientId (Just (audience user))) pass
+                return user
+
+getBuildMode :: Snap BuildMode
+getBuildMode = getParam "mode" >>= \ case
+    Just "blocklyXML" -> return (BuildMode "blocklyXML")
+
+funblockRoutes :: ClientId -> [(B.ByteString, Snap ())]
+funblockRoutes clientId =
+    [ ("floadProject",   loadProjectHandler clientId)
```

Comment:
This is extremely unfortunate.  Why should all the blocks UI stuff be so separate?

I feel like if I merge this change, that will be the day that Blocks starts to bitrot.  All of a sudden, it will share almost no code with the main site, and so its code will get further behind over time, until it's an unmaintainable mess.  I am not ready to give up on it yet.

Reply:
I agree with you on this but new file structure for collaboration would require this to happen. A solution would be to let even funblocks use that but then it will only create some unnecessary files. Any suggestions for this?

Reply:
It seems like most of the file manipulation should be the same.  The blocks UI should have no trouble supporting folders and such.  The only change should be disabling collaborative editing.  (In theory, that could eventually be implemented too; it would just require a different OT model and integration with Blockly!)  My hope would be that they would mostly share code, with a few buttons being disabled.  But I don't understand the functionality yet, so it's hard for me to propose a specific answer.